doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,433 @@
+to the record layer immediately, subject to congestion and flow control
+limitations.
+
+The write record layer will need a call equivalent to `writev(2)` in order to
+avoid copying data needlessly.  Each packet is defined and sent to the write
+record layer in fragments:
+
+```c
+struct quic_buffer_st {
+    unsigned char *buffer;
+    size_t length;
+};
+
+typedef struct quic_buffer_st QUIC_BUFFER;
+
+int ossl_quic_write_bytes(QUIC_STREAM *s, STACK_OF(QUIC_BUFFER));
+```
+
+The write record layer is responsible for combining QUIC packets into datagrams.
+
+### ACK Handling and Loss Detector
+
+1. When a packet is sent, the packetiser needs to inform the ACK Manager.
+2. When a packet is ACKed, inform packetiser so it can drop sent frames.
+3. When a packet is lost, inform packetiser to create retransmission packet(s).
+4. When a packet is discarded without ACK/loss, inform packetiser to clean up.
+
+```c
+__owur int ossl_quic_packet_sent(SSL_QUIC_TX_PACKETISER *tx,
+                                 OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_acked(SSL_QUIC_TX_PACKETISER *tx,
+                                  OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_lost(SSL_QUIC_TX_PACKETISER *tx,
+                                 OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_discarded(SSL_QUIC_TX_PACKETISER *tx,
+                                      OSSL_QUIC_PACKET *packet);
+```
+
+#### Notes
+
+| Name here | Name in ACK Manager |
+| --- | --- |
+| `ossl_quic_packet_sent` | `QUIC_ACKM_on_tx_ack_packet` |
+| `ossl_quic_packet_acked` | `on_acked` |
+| `ossl_quic_packet_lost` | `on_lost` |
+| `ossl_quic_packet_discarded` | `on_discarded` |
+
+Packets
+-------
+
+Packets formats are defined in [RFC 9000 17.1 Packet Formats].
+
+### Packet types
+
+QUIC supports a number of different packets.  The combination of packets of
+different types as per [RFC 9000 12.2 Coalescing Packets], is done by the
+record layer.
+
+#### Version Negotiation Packet
+
+Refer to [RFC 9000 17.2.1 Version Negotiation Packet].
+
+#### Initial Packet
+
+Refer to [RFC 9000 17.2.2 Initial Packet].
+
+#### Handshake Packet
+
+Refer to [RFC 9000 17.2.4 Handshake Packet].
+
+#### App Data 0-RTT Packet
+
+Refer to [RFC 9000 17.2.3 0-RTT].
+
+#### App Data 1-RTT Packet
+
+Refer to [RFC 9000 17.3.1 1-RTT].
+
+#### Retry Packet
+
+Refer to [RFC 9000 17.2.5 Retry Packet.
+
+Packetisation and Processing
+----------------------------
+
+### Application data frames
+
+The packetiser builds application data frames after requesting a specific
+amount of application data.  If insufficient data is available, or buffer
+boundaries prevent fulfilling the entire request, the stream send buffer module
+is free to return a smaller amount of data.
+
+```c
+int ossl_quic_ssb_request_data(QUIC_CONNECTION *conn, size_t ideal_length,
+                               const void **data, size_t *actual_length);
+```
```

Comment:
I'm confused what the difference is between this function and `ossl_quic_get_app_data` defined earlier in this document.

Why are we asking a connection object for stream data? Shouldn't we be asking a stream object for it?

Reply:
It is the same.  Just with better arguments.  I've removed this.